Add 2019 summary post (#66)
prestosql/prestosql.io: _posts/2020-01-01-2019-summary.md

---
layout: post
title:  "Presto in 2019: Year in Review"
author: Martin Traverso
excerpt_separator: <!--more-->
---

What a great year for the Presto community! We started with the year with the launch of the 
[Presto Software Foundation](https://prestosql.io/blog/2019/01/31/presto-software-foundation-launch.html), 
with the long term goal of ensuring the project remains collaborative, open and independent from 
any corporate interest, for years to come.

Since then, the community around Presto has grown and consolidated. We’ve seen contributions 
from more than 120 people across over 20 companies. Every week, 280 users and developers 
interact in the project’s [Slack channel](https://prestosql.io/slack.html.

With the collaboration of companies such as [Starburst](https://starburstdata.com), [Qubole](https://qubole.com), 
[Varada](https://varada.io), [Twitter](https://twitter.com), [ARM Treasure Data](https://www.treasuredata.com),
[Wix](https://wix.com), [Redhat](https://www.redhat.com), and the [Big Things community](https://www.meetup.com/Big-things-are-happening-here/),

> we ran several Presto summits across the world ([Tel Aviv](https://prestosql.io/blog/2019/05/03/Presto-Conference-Israel.html),

Reviewer: make it separate item list

[San Francisco](https://prestosql.io/blog/2019/05/17/Presto-Summit.html), [Tokyo](https://prestosql.io/blog/2019/07/11/report-for-presto-conference-tokyo.html), 

> [Bangalore](https://prestosql.io/blog/2019/09/05/Presto-Summit-Bangalore.html)) and New York) which 

Reviewer: and this a separate sentence after the list

brought together thousands of members from the larger Presto community.

The project has been more active than ever. We completed 28 releases comprised of more than 2850 
commits in over 1500 pull requests. Of course, that alone is not a good measure of progress, so 
let’s take a closer look at everything that went in.

<!--more-->

# Improvements

> ## Language features

Reviewer: title case

* Support for ``LATERAL`` in the context of an outer join.
* Support for ``UNNEST`` in the context of ``LEFT JOIN``. With this feature, it is now possible 
  to preserve the outer row when the array contains zero elements or is NULL.
	
  ```sql	
  SELECT * FROM t LEFT JOIN UNNEST(t.a) u(v) ON true
  ```

* Window functions now support the IGNORE NULLS clause. This is useful when combined with 
  functions such as LEAD, LAG, FIRST_VALUE, LAST_VALUE and NTH_VALUE if the dataset contains nulls.
* `.*` for ROW expressions.
* `ROW` types now support the subscript operator to access fields by index. This improves usability 
  when dealing with ROW type containing anonymous fields.
* [`CREATE SCHEMA`](https://prestosql.io/docs/current/sql/create-schema.html) syntax and support in various connectors.

## Query engine

* Extend support for conditional, lazy loading and processing (a.k.a., late materialization) to 
  Join, Window, TopN and Semijoin operators. This can dramatically reduce latency, CPU and I/O 
  for highly selective queries.
* [Unwrap cast/predicate pushdown](https://prestosql.io/blog/2019/05/21/optimizing-the-casts-away.html) optimizations.

> * Limit, table sample, projection pushdown into connectors.

Reviewer: Needs to be clarified a bit more

  and [`from_encoded_polyline`](https://prestosql.io/docs/current/functions/geospatial.html#from_encoded_polyline) functions.
* [`geometry_from_hadoop_shape`](https://prestosql.io/docs/current/functions/geospatial.html#geometry_from_hadoop_shape) to decode geometry objects in 
  Spatial Framework for Hadoop representation

## Security

* [Role-based access control](https://prestosql.io/docs/current/sql/create-role.html) and related commands.
* [INVOKER security mode](https://prestosql.io/docs/current/sql/create-view.html#security) for views, which allows views to be run using the permissions of the 
  current user.
* Secure results in client APIs.
* JWT-based [internal communication](https://prestosql.io/docs/current/security/internal-communication.html#internal-authentication) authentication.
* Credential passthrough, which allows Presto to authenticate with the underlying data source with 
  credentials provided by the user running a query.
* Impersonation for [Hive metastore](https://prestosql.io/docs/current/connector/hive.html#hive-thrift-metastore-configuration-properties).
* Support for Hadoop KMS.

> ## Cloud integration

Reviewer: title case

* Support for Azure Data Lake Blob and ADLS Gen2 storages
* Support for [Google Cloud Storage](https://prestosql.io/docs/current/connector/hive-gcs-tutorial.html).
* Several [performance improvements](https://github.com/prestosql/presto/pull/142) for AWS S3.

## CLI

* Syntax highlighting and multi-line editing.
 
  ![](/assets/blog/2019-review/presto-cli.jpeg)

* JSON output format and improvements to CSV output format.

## New Connectors

* [Elasticsearch](https://prestosql.io/docs/current/connector/elasticsearch.html)
* [Google Sheets](https://prestosql.io/docs/current/connector/googlesheets.html)
* [Amazon Kinesis](https://prestosql.io/docs/current/connector/kinesis.html)
* [Apache Phoenix](https://prestosql.io/blog/2019/06/04/phoenix-connector.html)
* [MemSQL](https://prestosql.io/docs/current/connector/memsql.html)
* Apache Iceberg

> ## Other

Reviewer: Other Improvements

  for Hive tables. This allows inserting data into bucketed tables without having to rewrite entire partitions.
* Lots and lots of bug fixes!

# Coming up...

These are some of the projects that are currently in progress and are likely to land in the short term.

* Support for pushing down row dereference expressions into connectors. This will help reduce 
  the amount of data and CPU needed to process highly nested columnar formats such as ORC and Parquet.
* Extend dynamic filtering to support distributed joins and other operators. Use dynamic filters for 
  pruning partitions at runtime when querying Hive.
* [Hive 3 support](https://prestosql.io/blog/2019/12/28/hive-3.html).
* [Improvements and features](https://github.com/prestosql/presto/issues/1324) in Iceberg connector.
* [Pinot](https://github.com/prestosql/presto/pull/2028) connector.
* [Oracle](https://github.com/prestosql/presto/pull/1959) connector.
* [Influx](https://github.com/prestosql/presto/pull/2397) connector.
* [Prometheus](https://github.com/prestosql/presto/pull/2321) connector.
* [Salesforce](https://prestosql.slack.com/archives/CQT2JH4KG/p1576038838027500) connector. 

> * [Prometheus]((https://github.com/prestosql/presto/pull/2321)) connector.

Reviewer: duplicate

---
layout: post
title:  "Presto in 2019: Year in Review"
author: Martin Traverso
excerpt_separator: <!--more-->
---

What a great year for the Presto community! We started with the year with the launch of the 
[Presto Software Foundation]({% post_url 2019-01-31-presto-software-foundation-launch %}), 
with the long term goal of ensuring the project remains collaborative, open and independent from 
any corporate interest, for years to come.

Since then, the community around Presto has grown and consolidated. We’ve seen contributions 
from more than 120 people across over 20 companies. Every week, 280 users and developers 
interact in the project’s [Slack channel](/slack.html).

With the collaboration of companies such as [Starburst](https://starburstdata.com), [Qubole](https://qubole.com), 
[Varada](https://varada.io), [Twitter](https://twitter.com), [ARM Treasure Data](https://www.treasuredata.com),

> [Wix](https://wix.com), [Redhat](https://www.redhat.com), and the [Big Things community](https://www.meetup.com/Big-things-are-happening-here/),

Reviewer 2: according to https://www.redhat.com/en:

Redhat -> Red Hat

we ran several Presto summits across the world ([Tel Aviv]({% post_url 2019-05-03-Presto-Conference-Israel %}),
[San Francisco]({% post_url 2019-05-17-Presto-Summit %}), [Tokyo]({% post_url 2019-07-11-report-for-presto-conference-tokyo %}), 
[Bangalore]({% post_url 2019-09-05-Presto-Summit-Bangalore %}) and New York) which 
brought together thousands of members from the larger Presto community.

The project has been more active than ever. We completed 28 releases comprised of more than 2850 
commits in over 1500 pull requests. Of course, that alone is not a good measure of progress, so 

> let’s take a closer look at everything that went in.

Reviewer 2: end the excerpt with something that gets the reader involved, eg

```suggestion
let’s take a closer look at everything that went in. And there is a lot to look at!
```

<!--more-->

# Improvements

> ## Language Features

Reviewer 3: support for correlated subqueries containing `LIMIT` or `TOP N`

> * [`FETCH FIRST n ROWS [ONLY] [WITH TIES]`](/docs/current/sql/select.html#limit-or-fetch-first-clauses) 

Reviewer 2: it's alternative

```suggestion
* [`FETCH FIRST n ROWS [ONLY | WITH TIES]`](/docs/current/sql/select.html#limit-or-fetch-first-clauses) 
```

standard syntax. The `WITH TIES` clause is particularly useful when some of the rows have the same 
value for the columns being used to order the results of a query.

> 	(TODO: sample query)

Reviewer 2: Let's use this example
```
SELECT student_name, score
FROM student s JOIN exam_result e ON s.id = e.student_id
ORDER BY score
FETCH FIRST 5 ROWS WITH TIES
```

> * [`OFFSET`](/docs/current/sql/select.html#offset-clause) syntax.

Reviewer 2: ```suggestion
* [`OFFSET`](/docs/current/sql/select.html#offset-clause) syntax, which is especially useful in ad-hoc queries,
```

* [`COMMENT ON <table>`](/docs/current/sql/comment.html) syntax to 
  set or remove table comments, and ability to see table comments via `DESCRIBE` or the 
  `system.metadata.table_comments` table.
* Support for ``LATERAL`` in the context of an outer join.
* Support for ``UNNEST`` in the context of ``LEFT JOIN``. With this feature, it is now possible 

>   to preserve the outer row when the array contains zero elements or is NULL.

Reviewer 2: ```suggestion
  to preserve the outer row when the array contains zero elements or is NULL.
  (We believe that majority of queries using `CROSS JOIN UNNEST` should actually be using `LEFT JOIN UNNEST`.)
```

	
  ```sql	
  SELECT * FROM t LEFT JOIN UNNEST(t.a) u(v) ON true
  ```

* Window functions now support the IGNORE NULLS clause. This is useful when combined with 
  functions such as LEAD, LAG, FIRST_VALUE, LAST_VALUE and NTH_VALUE if the dataset contains nulls.

> * `.*` for ROW expressions.

Reviewer 2: ```
* `ROW` expansion using `.*` operator.
```

> * `ROW` types now support the subscript operator to access fields by index. This improves usability 

Reviewer 2: ```suggestion
* `ROW` types now support the subscript operator to access fields by index. This greatly improves usability and readability of queries
```

  when dealing with ROW type containing anonymous fields.
  
    ![](/assets/blog/2019-review/row-ordinal.png)

* [`CREATE SCHEMA`](/docs/current/sql/create-schema.html) syntax and support 
  in various connectors (Hive, Iceberg, MySQL, PostgreSQL, Redshift, SQL Server, Phoenix).

## Query Engine

> * Extend support for conditional, lazy loading and processing (a.k.a., late materialization) to 

Reviewer 2: > Extend support for ... late materialization

It's not correct. We were using work processors in on 2019-01-01, but we were not using them for Late Materialization. They were used inside Order By and Window operators for implementing spill/no-spill paths.

So this should say
```
* Implement Late Materialization -- support for conditional and lazy loading of data during processing.
  Data is read from the storage only if it is actually needed, which may save a lot of CPU cycles (depending
  on the actual query predicates). This provide holistic and efficient processing of data from table scan through
  Join, Window, TopN and Semijoin operators. This not only improves CPU usage, but can also 
  dramatically reduce latency and I/O for selective queries.
```

Reviewer 2: "This is probably the most important performance optimization in years and we will be blogging about this more in coming weeks."

Author: That's why I phrased it that way. We currently support late materialization for Scan-Filter-Project. With the improvements to work processors last year, that is now expanded to other operators. I specifically did not talk about work processors -- that's just the under-the-covers implementation to enable generic late materialization.

Reviewer 2: I agree we should mention any technicalities here.

I didn't think that way. I though about Late Materialization (beyond Scan-Filter-Project) as something totally new.
And i think it's fair to phrase this as something new. Up to you.

Author: But we already have a form of late materialization (S-F-P), so saying "implement late materialization" is not accurate. Hence why I said "extend" (sure, via a completely new mechanism...)

Reviewer 2: Okay. So you can say "Generalize Late Materialization beyond Table Scan+Filter+Project. [...]"

> * Extend support for conditional, lazy loading and processing (a.k.a., late materialization) to 

Reviewer 3: * Better CBO support for projections:
https://github.com/prestosql/presto/pull/1431
https://github.com/prestosql/presto/pull/1359
* improved subquery planning (from performance point of view)
https://github.com/prestosql/presto/pull/1329
* improved reporting of physical data and network data read/sent
https://github.com/prestosql/presto/pull/326
https://github.com/prestosql/presto/pull/271
https://github.com/prestosql/presto/pull/133
* spill does not incur performance penalty when not triggered:
https://github.com/prestosql/presto/pull/315 
* spill for window and order by operators and spill productization in general:
https://github.com/prestosql/presto/pull/228

  Join, Window, TopN and Semijoin operators. This can dramatically reduce latency, CPU and I/O 
  for highly selective queries.
* [Unwrap cast/predicate pushdown]({% post_url 2019-05-21-optimizing-the-casts-away %}) optimizations.
* Limit, table sample, projection pushdown into connectors.

> * [Dynamic filtering]({% post_url 2019-06-30-dynamic-filtering %}) for broadcast joins.

Reviewer 2: ```suggestion
* [Dynamic filtering]({% post_url 2019-06-30-dynamic-filtering %}).
```

Author: Why?

Reviewer 2: It's not release notes. we don't need to be precise. I think this additional "for broadcast joins" simply dilutes the message.

## Performance

> * Improve performance of [`UNNEST`]({% post_url 2019-08-23-unnest-operator-performance-enhancements %}) by 10x.

Reviewer 2: ```suggestion
* Dramatic performance improvement of [`UNNEST`]({% post_url 2019-08-23-unnest-operator-performance-enhancements %}) by 10x.
```

Reviewer 2: Or 
```suggestion
* [10x performance improvement of `UNNEST`]({% post_url 2019-08-23-unnest-operator-performance-enhancements %}).
```

> * Improve performance of [ORC decoders]({% post_url 2019-04-23-even-faster-orc %}) by 2-7x resulting on a 

Reviewer 2: ```suggestion
* Huge improvement in performance of [ORC decoders]({% post_url 2019-04-23-even-faster-orc %}) by 2-7x resulting on a 
```

  10% global CPU improvement for the TPC-DS benchmark.

> * Improvements when reading small row groups in Parquet (e.g. small Parquet files, files with 

Reviewer 2: The braced expl is not necessary, but let's put this in context:

```suggestion
* Up to 17x speedup when reading small Parquet files or files with large number of columns, especially visible on S3 or other object storage.
```

17x -- this is a number from real production use-cases, i can describe this more.

  large number of columns, or files created by certain writers)
* Support for new ORC bloom filters
* Remove [redundant `ORDER BY`]({% post_url 2019-06-03-redundant-order-by %}) clauses.

> * Improvements for [`IN` / `NOT-IN`]({% post_url 2019-06-03-redundant-order-by %}) with subquery expressions (i.e., semijoin).

Reviewer 3: I would also mention:
* reduce query latency and Hive Metastore load (IMO it's really significant, but we didn't benchmark it):
https://github.com/prestosql/presto/pull/2184
https://github.com/prestosql/presto/pull/1921
https://github.com/prestosql/presto/pull/1866
* reduced query latency by smarter metadata handling (this is especially significant for JDBC connectors as for some of them querying `SHOW SCHEMAS/SHOW TABLES` took minutes. Now it only takes seconds)
https://github.com/prestosql/presto/pull/1455
https://github.com/prestosql/presto/pull/1329 (!)
https://github.com/prestosql/presto/pull/1308
https://github.com/prestosql/presto/pull/1306 (!)

> 

Reviewer 2: ```
* Make Cost-Based Optimizer (CBO) aware of estimated query peak memory footprint. This is
  especially useful for optimizing bigger queries, where not all parts of the query can be run
  concurrently.
```
https://github.com/prestosql/presto/pull/247

  [`with_timezone`](/docs/current/functions/datetime.html#with_timezone).

## Geospatial

* New geospatial functions: 
  [`ST_Points`](/docs/current/functions/geospatial.html#ST_Points), 
  [`ST_Length`](/docs/current/functions/geospatial.html#ST_Length), 
  [`ST_Area`](/docs/current/functions/geospatial.html#ST_Area), 
  [`line_interpolate_point`](/docs/current/functions/geospatial.html#line_interpolate_point) and 
  [`line_interpolate_points`](/docs/current/functions/geospatial.html#line_interpolate_points).
* `SphericalGeography` type and [related functions](/docs/current/functions/geospatial.html#to_spherical_geography) 
  to support spatial features in geographic coordinates (latitude / longitude) on spherical model of the earth.
* Support for Google Maps Polyline format via [`to_encoded_polyline`](/docs/current/functions/geospatial.html#to_encoded_polyline)
  and [`from_encoded_polyline`](/docs/current/functions/geospatial.html#from_encoded_polyline) functions.
* [`geometry_from_hadoop_shape`](/docs/current/functions/geospatial.html#geometry_from_hadoop_shape) to decode geometry objects in 
  Spatial Framework for Hadoop representation

> ## Security

Reviewer 2: move this above `Geospatial`

> * [Role-based access control](/docs/current/sql/create-role.html) and related commands.

Reviewer 3: I think it's 2018

* [INVOKER security mode](/docs/current/sql/create-view.html#security) for views, which allows views to be run using the permissions of the 
  current user.
* Prevent replay attacks and result hijacking in client APIs.

> * JWT-based [internal communication](/docs/current/security/internal-communication.html#internal-authentication) authentication.

Reviewer 2: ```suggestion
* JWT-based [internal communication](/docs/current/security/internal-communication.html#internal-authentication) authentication, which obsoletes the need to use Kerberos or certificates and greatly simplifies secure setups.
```

* Credential passthrough, which allows Presto to authenticate with the underlying data source with 

>   credentials provided by the user running a query.

Reviewer 2: ```suggestion
  credentials provided by the user running a query (we found this especially useful when dealing with
  Google Storage in GCP, but it's applied to a lot more use-cases as well).
```

* Impersonation for [Hive metastore](/docs/current/connector/hive.html#hive-thrift-metastore-configuration-properties).
* Support for reading and writing encrypted files in HDFS using Hadoop KMS.

## Cloud Integration

* Support for Azure Data Lake Blob and ADLS Gen2 storage.
* Support for [Google Cloud Storage](/docs/current/connector/hive-gcs-tutorial.html).
* Several [performance improvements](https://github.com/prestosql/presto/pull/142) for AWS S3.

## CLI

* Syntax highlighting and multi-line editing.
 
  ![](/assets/blog/2019-review/presto-cli.jpeg)

* JSON output format and improvements to CSV output format.

> 

Reviewer 2: ```
* Support & stability improvements for running CLI with Java 11.
```

This also applies to JDBC, so maybe call the section "CLI & JDBC", so we don't add a new JDBC section with just one bullet

Author: What do the "allow committing empty transaction" and "restoring autocommit mode" help with? Was it better support for some third-party tool? I prefer to describe in terms of the impact.

Reviewer 2: I didn't check with JDBC specs, but this makes it more compliant with other JDBC drivers and so it makes Presto JDBC more compatible with libraries.
(Specifically, i found this issues while trying to use Presto JDBC with a popular library)

## New Connectors

* [Elasticsearch](/docs/current/connector/elasticsearch.html)
* [Google Sheets](/docs/current/connector/googlesheets.html)
* [Amazon Kinesis](/docs/current/connector/kinesis.html)
* [Apache Phoenix]({% post_url 2019-06-04-phoenix-connector %})
* [MemSQL](/docs/current/connector/memsql.html)
* Apache Iceberg

> ## Other Improvements

Reviewer 2: Add
```
* New features and improvements in type mappings in PostgreSQL, MySQL, SQL Server and Redshift
  connectors. This includes support for PostgreSQL arrays and ability to read columns of unsupported 
  types.
```

Reviewer 2: Generally we could expand the connectors section with other examples too.

> * [Docker image](https://hub.docker.com/r/prestosql/presto) that provides an out-of-the-box single node 

Reviewer 2: ```suggestion
* [Presto Docker image](https://hub.docker.com/r/prestosql/presto) that provides an out-of-the-box single node 
```

  cluster with the JMX, memory, TPC-DS, and TPC-H catalogs. It can be deployed as a full cluster by 

>   mounting in configuration.

Reviewer 2: ```suggestion
  mounting in configuration and is being used for Kubernetes deployments.
```

* LZ4/ZSTD support for Parquet and ORC. These are the recommended compression algorithms.

> * Support for insert-only Hive transactional tables and bucketing v2 as part of 

Reviewer 2: ```suggestion
* Support for insert-only Hive transactional tables and Hive bucketing v2 as part of 
```

  [making Presto compatible with Hive 3]({% post_url 2019-12-28-hive-3 %}).
* Improvements in ANALYZE statement for Hive connector.
* Support for [multiple files per bucket]({% post_url 2019-05-29-improved-hive-bucketing %}) 
  for Hive tables. This allows inserting data into bucketed tables without having to rewrite entire partitions.
* Lots and lots of bug fixes!

# Coming Up...

These are some of the projects that are currently in progress and are likely to land in the short term.

> * Support for pushing down row dereference expressions into connectors. This will help reduce 

Reviewer 3: you might mention
* positions pushdown to reader (https://github.com/prestosql/presto/pull/2254)
* extended late materializaton support for complex subqueries (https://github.com/prestosql/presto/pull/2418)
* Improved insert to partitioned tables (https://github.com/prestosql/presto/pull/2358)

  the amount of data and CPU needed to process highly nested columnar formats such as ORC and Parquet.
* Extend dynamic filtering to support distributed joins and other operators. Use dynamic filters for 
  pruning partitions at runtime when querying Hive.

> * [Hive 3 support]({% post_url 2019-12-28-hive-3 %}).

Reviewer 2: ```suggestion
* full [Hive 3 support]({% post_url 2019-12-28-hive-3 %}).
```

* [Improvements and features](https://github.com/prestosql/presto/issues/1324) in Iceberg connector.
* [Pinot](https://github.com/prestosql/presto/pull/2028) connector.
* [Oracle](https://github.com/prestosql/presto/pull/1959) connector.
* [Influx](https://github.com/prestosql/presto/pull/2397) connector.
* [Prometheus](https://github.com/prestosql/presto/pull/2321) connector.
* [Salesforce](https://prestosql.slack.com/archives/CQT2JH4KG/p1576038838027500) connector. 
* Support for [Confluent registry in Kafka connector](https://github.com/prestosql/presto/pull/2106).
* Revamp of the function registry and function resolution to support dynamically-resolved 
  functions and SQL-defined functions.
* A new [Parquet writer](https://github.com/prestosql/presto/pull/2004) optimized to work efficiently 

>   within Presto.

Reviewer 2: ```
* ... and many, many more. Follow us on [Twitter](https://twitter.com/prestosql) and [join our slack](https://prestosql.io/slack.html).
```

[Varada](https://varada.io), [Twitter](https://twitter.com), [ARM Treasure Data](https://www.treasuredata.com),
[Wix](https://wix.com), [Red Hat](https://www.redhat.com), and the [Big Things community](https://www.meetup.com/Big-things-are-happening-here/),
we ran several Presto summits across the world: 
* [Tel Aviv, Israel, April 2019]({% post_url 2019-05-03-Presto-Conference-Israel %})
* [San Francisco, USA, June 2019]({% post_url 2019-05-17-Presto-Summit %})
* [Tokyo, Japan, July 2019]({% post_url 2019-07-11-report-for-presto-conference-tokyo %}) 
* [Bangalore, India, September, 2019]({% post_url 2019-09-05-Presto-Summit-Bangalore %})
* [New York, USA, December 2019](https://www.starburstdata.com/technical-blog/nyc-presto-summit-recap/) 

All these events were a huge success and brought thousands of Presto users, contributors and other community members together to 
share their knowledge.

The project has been more active than ever. We completed 28 releases comprised of more than 2850 
commits in over 1500 pull requests. Of course, that alone is not a good measure of progress, so 
let’s take a closer look at everything that went in. And there is a lot to look at!

> 

Author: I'm not sure whether it's possible to do that with Jekyll. @electrum ?

<!--more-->

# Improvements

## Language Features

* [`FETCH FIRST n ROWS [ONLY | WITH TIES]`](/docs/current/sql/select.html#limit-or-fetch-first-clauses) 
  standard syntax. The `WITH TIES` clause is particularly useful when some of the rows have the same 
  value for the columns being used to order the results of a query. Consider a case where you want to 
  list top 5 students with highest score on an exam. if 6th person has the same score as the 5th, you 
  want to know this as well, instead of getting an arbitrary and non-deterministic result:

  ```sql
  SELECT student_name, score
  FROM student s JOIN exam_result e USING (student_id)
  ORDER BY score
  FETCH FIRST 5 ROWS WITH TIES
  ```

* [`OFFSET`](/docs/current/sql/select.html#offset-clause) syntax, which is especially useful in ad-hoc queries.
* [`COMMENT ON <table>`](/docs/current/sql/comment.html) syntax to 

>   set or remove table comments, and ability to see table comments via `DESCRIBE` or the 

Reviewer 4: We've shown comments with `DESCRIBE` forever. How about making this two sentences
```
* [`COMMENT ON <table>`](/docs/current/sql/comment.html) syntax to 
  set or remove table comments. Comments can be shown via `DESCRIBE`
  or the new `system.metadata.table_comments` table.

  `system.metadata.table_comments` table.

> * Support for ``LATERAL`` in the context of an outer join.

Reviewer 4: This is markdown, so single backtick

> * Support for ``UNNEST`` in the context of ``LEFT JOIN``. With this feature, it is now possible 

Author: No reason at all. Just bad habits from writing `rst` release notes. I'll fix them all.

  to preserve the outer row when the array contains zero elements or is NULL. Most common usages
  of `UNNEST` in a `CROSS JOIN` should actually be using this form.
	
  ```sql	

>   SELECT * FROM t LEFT JOIN UNNEST(t.a) u(v) ON true

Reviewer 4: Add space between table name `u` and column list.

  ```

> * Window functions now support the IGNORE NULLS clause. This is useful when combined with 

Reviewer 4: Code quotes on `IGNORE NULLS`

Reviewer 4: Maybe invert this, so it matches the flow of other comments
```
`IGNORE NULLS` clause for window functions.

>   functions such as LEAD, LAG, FIRST_VALUE, LAST_VALUE and NTH_VALUE if the dataset contains nulls.

Reviewer 4: Make these function names lowercase but use code quotes
```
  functions such as `lead`, `lag`, `first_value`, `last_value` and `nth_value` if the dataset contains nulls.

* `ROW` expansion using `.*` operator.

> * `ROW` types now support the subscript operator to access fields by index. This This greatly improves usability 

Reviewer 4: ```
Subscript operator to access `ROW` type fields by index.
```

> * `ROW` types now support the subscript operator to access fields by index. This This greatly improves usability 

Reviewer 4: Duplicate word "This"

>   and readability of queries when dealing with ROW type containing anonymous fields.

Reviewer 4: Add code quotes or lowercase

Reviewer 4: Grammar: should be "row types"

Reviewer 4: We might want to put this item last, since the large image splits off the last two items in the list

  
    ![](/assets/blog/2019-review/row-ordinal.png)

* [`CREATE SCHEMA`](/docs/current/sql/create-schema.html) syntax and support 
  in various connectors (Hive, Iceberg, MySQL, PostgreSQL, Redshift, SQL Server, Phoenix).
* Support for correlated subqueries containing `LIMIT` or `ORDER BY`+`LIMIT`.

## Query Engine

* Generalize conditional, lazy loading and processing (a.k.a., late materialization) beyond 

>   Table Scan, Filter and Projection to support Join, Window, TopN and Semijoin operators. This can dramatically 

Reviewer 4: Should it be SemiJoin since it seems like two words?

  reduce latency, CPU and I/O for highly selective queries. This is one of the most important performance 
  optimizations in recent times and we will be blogging about this more in coming weeks.
* [Unwrap cast/predicate pushdown]({% post_url 2019-05-21-optimizing-the-casts-away %}) optimizations.

> * Add support for pushing down operations such as limit, table sample, or projections into connectors during

Reviewer 4: We should avoid "add support for" here since this is a list of features (nouns).

> Connector pushdown during planning for operations such as limit, table sample, or projections.

Author: I don't see what's wrong with that form, but I'll change it.

  query planning. This allows connectors to optimize how data is accessed before it's provided to the 
  Presto engine for further processing. 
* [Dynamic filtering]({% post_url 2019-06-30-dynamic-filtering %}).
* Cost-Based Optimizer can now consider [estimated query peak memory](https://github.com/prestosql/presto/pull/247) 
  footprint. This is especially useful for optimizing bigger queries, where not all parts of the query can 
  be run concurrently.
* Improved handling of [projections](https://github.com/prestosql/presto/pull/1431), 

>   [aggregations](https://github.com/prestosql/presto/pull/864) and [cross joins](https://github.com/prestosql/presto/pull/1359) 

Author: That's the one @sopel39 linked to...

  in cost based optimizer.
* Improved accounting and reporting of physical and network data read or transmitted during query processing.    

## Performance

* [10x performance improvement for `UNNEST`]({% post_url 2019-08-23-unnest-operator-performance-enhancements %}).
* 2-7x improvement in performance of [ORC decoders]({% post_url 2019-04-23-even-faster-orc %}), resulting in a 
  10% global CPU improvement for the TPC-DS benchmark.
* Improvements when reading small Parquet files, files with large number of columns, or files with small row
  groups.
* Support for new ORC bloom filters.
* Remove [redundant `ORDER BY`]({% post_url 2019-06-03-redundant-order-by %}) clauses.
* Improvements for [`IN` / `NOT-IN`]({% post_url 2019-06-03-redundant-order-by %}) with subquery expressions (i.e., semijoin).
* Huge performance improvements when [reading from `information_schema`](https://github.com/prestosql/presto/pull/1329).

> * Reduce query latency and Hive Metastore load.

Reviewer 4: Lowercase metastore, since we don't capitalize it elsewhere

* Improve metadata handling during planning. This can result in dramatic improvements in latency, 
  especially for connectors such as MySQL, PostgreSQL, Redshift, SQL Server, etc. Some queries like 

>   `SHOW SCHEMAS`/`SHOW TABLES` that could take several minutes to complete now finish in a few seconds.

Reviewer 4: Replace slash with "or"

> * Improved stability and performance when spilling is enabled.  

Author: > shall we mention spill encryption?

Yes!

> * Improved stability and performance when spilling is enabled.  

Reviewer 2: encryption went in together with compression

## Functions

* [`combinations`](/docs/current/functions/array.html#combinations)
* [`format`](/docs/current/functions/conversion.html#format)
* [`UUID` type](/docs/current/functions/uuid.html) and related functions.
* [`all_match`](/docs/current/functions/array.html#all_match),
  [`any_match`](/docs/current/functions/array.html#any_match) and 
  [`none_match`](/docs/current/functions/array.html#none_match).
* Support for higher-order [`reduce_agg`](/docs/current/functions/aggregate.html#reduce_agg) 

>   function in aggregation queries.

Reviewer 2: maybe instead of "higher-order" use wording easier to digest:

```
* Support flexible aggregation with lambdas using
  [`reduce_agg`](/docs/current/functions/aggregate.html#reduce_agg).
```

> * New datetime functions: [`last_day_of_month`](/docs/current/functions/datetime.html#last_day_of_month),

Reviewer 4: Let's match the wording in the manual

> New date and time functions

  [`at_timezone`](/docs/current/functions/datetime.html#at_timezone) and 
  [`with_timezone`](/docs/current/functions/datetime.html#with_timezone).

## Security

* [Role-based access control](/docs/current/sql/create-role.html) and related commands.
* [INVOKER security mode](/docs/current/sql/create-view.html#security) for views, which allows views to be run using the permissions of the 
  current user.
* Prevent replay attacks and result hijacking in client APIs.
* JWT-based [internal communication](/docs/current/security/internal-communication.html#internal-authentication) authentication,
  which obsoletes the need to use Kerberos or certificates and greatly simplifies secure setups.
* Credential passthrough, which allows Presto to authenticate with the underlying data source with 

>   credentials provided by the user running a query. We found this especially useful when dealing with

Reviewer 4: I'd remove "we found" to better match the style of the rest of the post and replace with

> This is especially useful when ...

  Google Storage in GCP or SQL databases that manage user authentication and authorization on 
  their own.
* Impersonation for [Hive metastore](/docs/current/connector/hive.html#hive-thrift-metastore-configuration-properties).
* Support for reading and writing encrypted files in HDFS using Hadoop KMS.

## Geospatial

* New geospatial functions: 
  [`ST_Points`](/docs/current/functions/geospatial.html#ST_Points), 
  [`ST_Length`](/docs/current/functions/geospatial.html#ST_Length), 
  [`ST_Area`](/docs/current/functions/geospatial.html#ST_Area), 
  [`line_interpolate_point`](/docs/current/functions/geospatial.html#line_interpolate_point) and 
  [`line_interpolate_points`](/docs/current/functions/geospatial.html#line_interpolate_points).
* `SphericalGeography` type and [related functions](/docs/current/functions/geospatial.html#to_spherical_geography) 

>   to support spatial features in geographic coordinates (latitude / longitude) on spherical model of the earth.

Reviewer 4: "on spherical model" sounds wrong grammatically. How about "using a spherical model"

Author: That was meant to say "on a spherical model"

* Support for Google Maps Polyline format via [`to_encoded_polyline`](/docs/current/functions/geospatial.html#to_encoded_polyline)
  and [`from_encoded_polyline`](/docs/current/functions/geospatial.html#from_encoded_polyline) functions.
* [`geometry_from_hadoop_shape`](/docs/current/functions/geospatial.html#geometry_from_hadoop_shape) to decode geometry objects in 
  Spatial Framework for Hadoop representation

## Cloud Integration

* Support for Azure Data Lake Blob and ADLS Gen2 storage.
* Support for [Google Cloud Storage](/docs/current/connector/hive-gcs-tutorial.html).

> * Several [performance improvements](https://github.com/prestosql/presto/pull/142) for AWS S3.

Reviewer 4: Link to blog post
```
{% post_url 2019-05-06-faster-s3-reads %}

## CLI and JDBC Driver

* Syntax highlighting and multi-line editing.
 

>   ![](/assets/blog/2019-review/presto-cli.jpeg)

Reviewer 4: The image has a shaded gray border around it. Let's trim that off so the image is just the CLI window.

* JSON output format and improvements to CSV output format.

> * Support & stability improvements for running the CLI and JDBC driver with Java 11.

Reviewer 4: Use "and" instead of &

* Improve compatibility of JDBC driver with third-party tools.

## New Connectors

* [Elasticsearch](/docs/current/connector/elasticsearch.html)
* [Google Sheets](/docs/current/connector/googlesheets.html)
* [Amazon Kinesis](/docs/current/connector/kinesis.html)
* [Apache Phoenix]({% post_url 2019-06-04-phoenix-connector %})
* [MemSQL](/docs/current/connector/memsql.html)

> * Apache Iceberg

Reviewer 4: Let's add a disclaimer, something like
```
* Apache Iceberg (preview version still under development)

> ## Other Improvements

Reviewer 4: Add connector view definitions (might be too detailed)

* Connector provided view definitions, allowing the connector to dynamically generate the definition at query time. For example, the connector can provide a union of two tables, filtered on a disjoint time range, with the cutoff time determined at resolution time.

* [Presto Docker image](https://hub.docker.com/r/prestosql/presto) that provides an out-of-the-box single node 
  cluster with the JMX, memory, TPC-DS, and TPC-H catalogs. It can be deployed as a full cluster by 
  mounting in configuration and can be used for Kubernetes deployments.

> * LZ4/ZSTD support for Parquet and ORC. These are the recommended compression algorithms.

Reviewer 2: ```suggestion
* Support LZ4 and ZSTD compression codecs in Parquet and ORC. These are the recommended compression algorithms (LZ4 for fast lightweight compression and ZSTD otherwise).
```

> * LZ4/ZSTD support for Parquet and ORC. These are the recommended compression algorithms.

Reviewer 4: Use "and" instead of slash since these are two separate algorithms. Also, I think it's written as "Zstd"

* Support for insert-only Hive transactional tables and Hive bucketing v2 as part of 
  [making Presto compatible with Hive 3]({% post_url 2019-12-28-hive-3 %}).

> * Improvements in ANALYZE statement for Hive connector.

Reviewer 4: Code quotes for ANALYZE

* Support for [multiple files per bucket]({% post_url 2019-05-29-improved-hive-bucketing %}) 

>   for Hive tables. This allows inserting data into bucketed tables without having to rewrite entire partitions.

Reviewer 2: ```suggestion
  for Hive tables. This allows inserting data into bucketed tables without having to rewrite entire partitions
  and improves Presto compatibility with Hive and other tools.
```

* Support for upper- and mixed-case table and column names in JDBC-based connectors.
* New features and improvements in type mappings in PostgreSQL, MySQL, SQL Server and Redshift
  connectors. This includes support for PostgreSQL arrays and `timestamp with time zone` type, and 
  the ability to read columns of unsupported types.
* Improvements in [Hive compatibility with Hive version 2.3](https://github.com/prestosql/presto/pull/833) 

>   and [with CDH's Hive](https://github.com/prestosql/presto/pull/1937).

Reviewer 2: ```suggestion
  and [with Cloudera (CDH)'s Hive](https://github.com/prestosql/presto/pull/1937).
```

??

* Lots and lots of bug fixes!

# Coming Up...

These are some of the projects that are currently in progress and are likely to land in the short term.

* Support for pushing down row dereference expressions into connectors. This will help reduce 
  the amount of data and CPU needed to process highly nested columnar formats such as ORC and Parquet.
* Extend dynamic filtering to support distributed joins and other operators. Use dynamic filters for 
  pruning partitions at runtime when querying Hive.

> * [Extended late materialization](https://github.com/prestosql/presto/pull/2418) support to queries involving 

Reviewer 2: ```suggestion
* [Extended Late Materialization](https://github.com/prestosql/presto/pull/2418) support to queries involving 
```

  complex correlated subqueries.
* Finalize [Hive 3 support]({% post_url 2019-12-28-hive-3 %}).
* Improved [INSERT into partitioned tables](https://github.com/prestosql/presto/pull/2358), which will help with 
  large ETL queries.
* [Improvements and features](https://github.com/prestosql/presto/issues/1324) in Iceberg connector.
* [Pinot](https://github.com/prestosql/presto/pull/2028) connector.
* [Oracle](https://github.com/prestosql/presto/pull/1959) connector.
* [Influx](https://github.com/prestosql/presto/pull/2397) connector.
* [Prometheus](https://github.com/prestosql/presto/pull/2321) connector.
* [Salesforce](https://prestosql.slack.com/archives/CQT2JH4KG/p1576038838027500) connector. 
* Support for [Confluent registry in Kafka connector](https://github.com/prestosql/presto/pull/2106).
* Revamp of the function registry and function resolution to support dynamically-resolved 
  functions and SQL-defined functions.
* A new [Parquet writer](https://github.com/prestosql/presto/pull/2004) optimized to work efficiently 
  within Presto.

... and many, many more. Follow us on [Twitter](https://twitter.com/prestosql) and 

> [join our Slack channel](https://prestosql.io/slack.html).

Reviewer 4: Remove "channel" since we are asking them to join our Slack instance (not sure what to call it but "channel" is wrong)

Reviewer 2: in slack parlance this is called "workspace", but i dont think we should refer to it as this.
+1 for "Slack" alone
